### PR TITLE
WT-6111 review cell structures

### PR DESCRIPTION
**Pull request by keithbostic**
Alex, this should give you an idea of where I think this branch is likely to end up.

The unpacked cell now comes in 3 flavors: `WT_CELL_UNPACK`, which is has both the `WT_TIME_WINDOW` and the `WT_TIME_AGGREGATE` structures, `WT_CELL_UNPACK_ADDR` which has only the `WT_TIME_AGGREGATE` structure, and `WT_CELL_UNPACK_VALUE` which has only the `WT_TIME_WINDOW` structure.

If you use the wrong time window fields you'll get either a compile error (if you use one of the limited versions of the unpacked cell, that is, `WT_CELL_UNPACK_ADDR` or `WT_CELL_UNPACK_VALUE`), and a core dump if you use the combined version of the unpacked cell, `WT_CELL_UNPACK`.

The changes to `cell.i` and `cell.h` are the interesting ones: I'd like a cleaner solution in `cell.h` but I don't really see one.

I've only implemented the limited versions of the unpacked cell in `btree/bt_page.c`, but that should give you an idea of what we'd do in other places like reconciliation.

It would be great to get rid of the unlimited version of the unpacked cell (`WT_CELL_UNPACK`), entirely, but that would mean splitting up the verify and debug functions (among others), that handle both address and key/value cells in one function, and that's a hill I'm not prepared to climb.

### CODE REVIEW

**Comment by keithbostic** [comment]
evergreen retry

**Comment by keithbostic** [comment]
evergreen retry

**Comment by keithbostic** [comment]
evergreen retry

**Comment by agorrod** [comment]
Thanks Keith, to be honest this isn't as clean as I'd hoped. Passing void * in some places and define the shared portion of the cell structure make the code less clean. I understand there are trade offs in each of those decisions.
Perhaps we can chat live to talk this one through?

**Comment by keithbostic** [comment]
evergreen retry

**Comment by keithbostic** [comment]
evergreen: retry

**Comment by keithbostic** [comment]
evergreen retry

File: src/btree/bt_ovfl.c [hunk 1]
```diff
@@ -65,7 +65,6 @@ __wt_ovfl_read(
      */
     __wt_readlock(session, &S2BT(session)->ovfl_lock);
     if (__wt_cell_type_raw(unpack->cell) == WT_CELL_VALUE_OVFL_RM) {
-        WT_ASSERT(session, __wt_txn_visible_all(session, unpack->tw.stop_txn, unpack->tw.stop_ts));
```
Comment (agorrod) [inline src/btree/bt_ovfl.c:68]: Oh -curious. Do overflow cells not have a time window?

File: src/conn/conn_open.c [hunk 2]
```diff
@@ -210,9 +210,18 @@ __wt_connection_workers(WT_SESSION_IMPL *session, const char *cfg[])
     /*
      * Run recovery. NOTE: This call will start (and stop) eviction if recovery is required.
      * Recovery must run before the history store table is created (because recovery will update the
-     * metadata), and before eviction is started for real.
+     * metadata, and set the maximum file id seen), and before eviction is started for real.
```
Comment (agorrod) [inline src/conn/conn_open.c:213]: I'm surprised this change is here - was it a merge conflict issue?

File: src/txn/txn_recover.c [hunk 3]
```diff
@@ -552,27 +553,6 @@ __wt_txn_recover(WT_SESSION_IMPL *session)
     metafile = &r.files[WT_METAFILE_ID];
     metafile->c = metac;
 
-    /*
```
Comment (agorrod) [inline src/txn/txn_recover.c:555]: Similarly to above - why is this change here? It looks orthogonal.

**Comment by agorrod** [comment]
@quchenhao could you do a review of this change please? It's an extension of the change in WT-6086 you reviewed yesterday.

File: src/btree/bt_vrfy_dsk.c [hunk 4]
```diff
@@ -858,7 +1031,7 @@ __verify_dsk_chunk(
     end = (uint8_t *)dsk + dsk->mem_size;
 
     /*
-     * Fixed-length column-store and overflow pages are simple chunks of data. Verify the data
+     * Fixed-length column-store and overflow pages are simple chunks of data-> Verify the data
```
Comment (quchenhao) [inline src/btree/bt_vrfy_dsk.c:1034]: Is this change by accident?

File: src/include/cell.i [hunk 5]
```diff
@@ -706,19 +710,32 @@ __wt_cell_leaf_value_parse(WT_PAGE *page, WT_CELL *cell)
...
+    } while (0)
+
 /*
  * __wt_cell_unpack_safe --
  *     Unpack a WT_CELL into a structure, with optional boundary checks.
  */
 static inline int
 __wt_cell_unpack_safe(WT_SESSION_IMPL *session, const WT_PAGE_HEADER *dsk, WT_CELL *cell,
-  WT_CELL_UNPACK *unpack, const void *end)
+  WT_CELL_UNPACK_ADDR *unpack_addr, WT_CELL_UNPACK_KV *unpack_value, const void *end)
```
Comment (quchenhao) [inline src/include/cell.i:731]: Can we break this into two functions? We may be accidentally passing both `unpack_addr` and `unpack_value` to this. To share common code between the two functions, we can share them through small inline functions.

File: src/include/cell.h [hunk 6]
```diff
@@ -144,37 +144,65 @@ struct __wt_cell {
...
+                                                /* AUTOMATIC FLAG VALUE GENERATION STOP */
+
 /*
- * WT_CELL_UNPACK --
- *	Unpacked cell.
+ * We have three flavors of cell unpack structures: one that can unpack any kind of cell, one that
+ * can only unpack address cells and one that can only unpack value cells. The reason is because we
+ * want to catch code using the wrong time window information relative to the cell type, minimize
+ * the size of the unpack structure, and finally reduce the code executed in performance paths,
+ * while still allowing other code such as verification and debug functions to be able to read any
```
Comment (quchenhao) [inline src/include/cell.h:158]: Does verification and debug functions know what type of cells they are reading?

**Comment by keithbostic** [comment]
@agorrod, @quchenhao, yeah, there are merge issues. I repointed the branch at develop, and that doesn't really help. I'm going to reply to your comments and then close down this branch, I don't think we can trust it.

File: src/btree/bt_ovfl.c (hunk 1 above)
Comment (keithbostic) [reply src/btree/bt_ovfl.c:68]: No; this code is the one (?) place where we ended up taking cells of both `WT_CELL_UNPACK_ADDR` and `WT_CELL_UNPACK_KV` types, and so we don't have a `WT_TIME_WINDOW`. Since it's just an assert, I removed it.

File: src/conn/conn_open.c (hunk 2 above)
Comment (keithbostic) [reply src/conn/conn_open.c]: Yeah. The merge is a mess.

File: src/txn/txn_recover.c (hunk 3 above)
Comment (keithbostic) [reply src/txn/txn_recover.c]: Yeah. The merge is a mess.

File: src/btree/bt_vrfy_dsk.c (hunk 4 above)
Comment (keithbostic) [reply src/btree/bt_vrfy_dsk.c:1034]: The merge is a mess.

File: src/include/cell.i (hunk 5 above)
Comment (keithbostic) [reply src/include/cell.i:731]: No. There is a single function that cracks a cell, and now that we crack cells into different structures, this function has to take references to both types. One of the arguments is always NULL, that is, we're always working with one type or another.

Obviously, we could split this function into two parts, but that's not something I'm eager do.

File: src/include/cell.h (hunk 6 above)
Comment (keithbostic) [reply src/include/cell.h:158]: Yes -- you're correct, this was a comment left over from an earlier version of this code where the debug & verification code hadn't been fixed.

**Comment by keithbostic** [comment]
Closing the PR, the merge can't be trusted.
